src/versioned_file.ml

```diff
@@ -0,0 +1,88 @@
+open Import
+
+module type S = sig
+  type data
+
+  module Lang : sig
+    val register : Syntax.t -> data -> unit
+    module Instance : sig
+      type t =
+        { syntax  : Syntax.t
+        ; data    : data
+        ; version : Syntax.Version.t
+        }
+    end
+    val get_exn : string -> Instance.t
+  end
+  val load : Path.t -> f:(Lang.Instance.t -> 'a Sexp.Of_sexp.t) -> 'a
+  val parse_contents
+    :  Lexing.lexbuf
+    -> Dune_lexer.first_line
+    -> f:(Lang.Instance.t -> 'a Sexp.Of_sexp.t)
+    -> 'a
+end
+
+module Make(Data : sig type t end) = struct
```

Comment:
Wondering what application you have in mind for this data field. It took me more time than I will admit that this argument isn't actually used anywhere. Would be simpler to just keep `Make` generative for now.